src/debug_info.rs

```diff
@@ -1,30 +1,1060 @@
+    /// is incomplete.  Module/function verification can then fail.
+    ///
+    /// Call before any kind of code generation (including verification). Can be called more than
+    /// one time.
+    pub fn finalize(&self) {
+        use llvm_sys::debuginfo::LLVMDIBuilderFinalize;
+        unsafe { LLVMDIBuilderFinalize(self.builder) };
     }
 }
 
-impl Drop for DebugInfoBuilder {
+impl<'ctx> Drop for DebugInfoBuilder<'ctx> {
     fn drop(&mut self) {
-        // FIXME: Docs say DIB must be finalize before calling this
+        use llvm_sys::debuginfo::LLVMDisposeDIBuilder;
+
+        self.finalize();
+        unsafe { LLVMDisposeDIBuilder(self.builder) }
+    }
+}
+
+/// Source file scope for debug info
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub struct DIFile<'ctx> {
+    pub(crate) metadata_ref: LLVMMetadataRef,
+    _marker: PhantomData<&'ctx Context>,
+}
+
+impl<'ctx> AsDIScope<'ctx> for DIFile<'ctx> {
+    fn as_debug_info_scope(self) -> DIScope<'ctx> {
+        DIScope {
+            metadata_ref: self.metadata_ref,
+            _marker: PhantomData,
+        }
+    }
+}
+
+/// Compilation unit scope for debug info
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub struct DICompileUnit<'ctx> {
+    file: DIFile<'ctx>,
+    pub(crate) metadata_ref: LLVMMetadataRef,
+    _marker: PhantomData<&'ctx Context>,
+}
+
+impl<'ctx> DICompileUnit<'ctx> {
+    pub fn get_file(&self) -> DIFile<'ctx> {
+        self.file
+    }
+}
+
+impl<'ctx> AsDIScope<'ctx> for DICompileUnit<'ctx> {
+    fn as_debug_info_scope(self) -> DIScope<'ctx> {
+        DIScope {
+            metadata_ref: self.metadata_ref,
+            _marker: PhantomData,
+        }
+    }
+}
+
+/// Function body scope for debug info
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub struct DISubProgram<'ctx> {
+    pub(crate) metadata_ref: LLVMMetadataRef,
+    pub(crate) _marker: PhantomData<&'ctx Context>,
+}
+
+impl<'ctx> AsDIScope<'ctx> for DISubProgram<'ctx> {
+    fn as_debug_info_scope(self) -> DIScope<'ctx> {
+        DIScope {
+            metadata_ref: self.metadata_ref,
+            _marker: PhantomData,
+        }
+    }
+}
+
+/// Any kind of debug info type
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub struct DIType<'ctx> {
+    pub(crate) metadata_ref: LLVMMetadataRef,
+    _marker: PhantomData<&'ctx Context>,
+}
+
+impl<'ctx> DIType<'ctx> {
+    pub fn get_size_in_bits(&self) -> u64 {
+        use llvm_sys::debuginfo::LLVMDITypeGetSizeInBits;
+        unsafe { LLVMDITypeGetSizeInBits(self.metadata_ref) }
+    }
+
+    pub fn get_align_in_bits(&self) -> u32 {
+        use llvm_sys::debuginfo::LLVMDITypeGetAlignInBits;
+        unsafe { LLVMDITypeGetAlignInBits(self.metadata_ref) }
+    }
+
+    pub fn get_offset_in_bits(&self) -> u64 {
+        use llvm_sys::debuginfo::LLVMDITypeGetOffsetInBits;
+        unsafe { LLVMDITypeGetOffsetInBits(self.metadata_ref) }
+    }
+}
+
+/// A debug info typedef, created by `create_typedef` method of `DebugInfoBuilder`
```

Comment:
`DIDerivedType` includes "qualified types, pointers, references, friends, typedefs, and class members".